### DISCUSSION TITLE
Correct typing for function taking keyword-only arguments and kwargs

### DISCUSSION THREAD

**Discussion by jrast** [python/typing#1406] [Q&A]
Hi! I have a question regarding the typing of a function with keyword-only arguments:
I have the following function:
```python
def send_command(name: str, *, timeout: Optional[float] = None, **params: Any):
    print(name, params, timeout)
```

Which is called like this:
```python
# Valid Calls
kwargs = {"p1": "abc", "timeout": 1.0}
send_command("Test", **kwargs)
send_command("Test", a=1, b=2, timeout=10)
# Invalid Calls
send_command("Test", timeout="not a number")  # Wrong type for timeout
```
With the typehints as shown above, mypy fails on the first call with `Argument 2 to "send_command" has incompatible type "**Dict[str, object]"; expected "Optional[float]"  [arg-type]`

I tried to fix this by adding the following overloads:
```python
@overload
def send_command(name: str, *, timeout: Optional[float], **params: Any):
    ...

@overload
def send_command(name: str, **params: Any):
    ...
```
However, with those overloads, mypy no longer detects the wrong type for the timeout parameter (third call from above). What would be the correct typehint for this function?

**Comment by srittau** [comment]
Unfortunately, mypy is not great at deriving the best type for the dict literal in this case. Try explicitly annotating the dict:

```python
kwargs: dict[str, Any] = {"p1": "abc", "timeout": 1.0}
```

**Reply by jrast** [reply]
I wanted to avoid this... The function is used all over the place.

But for me one question remains: Why does mypy report this as a error on the `timeout` parameter? For me it seems like mypy expects that the order is relevant in this case, which is not. For example `send_command("Test", **{"p1": "abc"})` also fails with the same error, see [mypy Playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=e55d95ddf8b276112d33c1aec2654a6c).

**Reply by erictraut** [reply]
This appears to be a bug in mypy. I recommend reporting it in the mypy issue tracker. It type checks as you would expect with pyright.

**Reply by jrast** [reply]
Created a bug report here: https://github.com/python/mypy/issues/15317
